sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/InputStreamContent.java

```diff
@@ -77,20 +95,95 @@ public ByteBuffer toByteBuffer() {
 
     @Override
     public Flux<ByteBuffer> toFluxByteBuffer() {
-        return FluxUtil.toFluxByteBuffer(this.content, STREAM_READ_SIZE);
+        return FluxUtil.toFluxByteBuffer(this.content.get(), STREAM_READ_SIZE);
     }
 
     @Override
     public boolean isReplayable() {
-        return false;
+        return isReplayable;
+    }
+
+    @Override
+    public BinaryDataContent toReplayableContent() {
+        if (isReplayable) {
+            return this;
+        } else {
+            InputStream inputStream = this.content.get();
+            if (canMarkReset(inputStream, length)) {
+                return createMarkResetContent(inputStream, length);
+            } else {
+                return readAndBuffer(inputStream, length);
+            }
+        }
+    }
+
+    @Override
+    public Mono<BinaryDataContent> toReplayableContentAsync() {
+        if (isReplayable) {
+            return Mono.just(this);
+        } else {
+            InputStream inputStream = this.content.get();
+            if (canMarkReset(inputStream, length)) {
+                return Mono.fromCallable(() -> createMarkResetContent(inputStream, length));
+            } else {
+                return Mono.just(inputStream)
+                    .publishOn(Schedulers.boundedElastic()) // reading stream can be blocking.
+                    .map(ignore -> readAndBuffer(inputStream, length));
+            }
+        }
+    }
+
+    private static boolean canMarkReset(InputStream inputStream, Long length) {
+        return length != null && length < Integer.MAX_VALUE - 8 && inputStream.markSupported();
+    }
+
+    private static InputStreamContent createMarkResetContent(InputStream inputStream, Long length) {
+        inputStream.mark(length.intValue());
+        return new InputStreamContent(
+            () -> {
+                try {
+                    inputStream.reset();
+                    return inputStream;
+                } catch (IOException e) {
+                    throw LOGGER.logExceptionAsError(new UncheckedIOException(e));
+                }
+            }, length, true
+        );
+    }
+
+    private static InputStreamContent readAndBuffer(InputStream inputStream, Long length) {
+        try {
+            Vector<ByteArrayInputStream> chunkInputStreams = new Vector<>();
+
+            int read;
+            do {
+                byte[] chunk = new byte[BUFFER_CHUNK_SIZE];
+                read = inputStream.read(chunk);
+                if (read > 0) {
+                    chunkInputStreams.add(new ByteArrayInputStream(chunk, 0, read));
+                }
+            } while (read >= 0);
+
+            return new InputStreamContent(
+                () -> {
+                    for (ByteArrayInputStream chunkInputStream : chunkInputStreams) {
+                        chunkInputStream.reset();
+                    }
+                    return new SequenceInputStream(chunkInputStreams.elements());
```

Comment:
Is the `Enumeration` replayable after the first consumption?

Reply:
Looking at `SequenceInputStream` this looks like no

<img width="1087" alt="image" src="https://user-images.githubusercontent.com/48699787/175400298-74696ec2-b108-4f6e-b4b5-02e2758e5f3b.png">

We may want to create our own internal variant that composes a list of byte array, will reduce some overhead too.

Reply:
```
BinaryData data = getStreamData().toReplayableBinaryData();
consume(data.toStream());
consume(data.toStream();
```

is going to work. the toStream calls the supplier which creates brand new enumeration.

Reply:
This doesn't call buffer again does it?